<?xml version="1.0"?>
<launch>
    
    <arg name="bag_file"                default="2023-05-08-14-46-13_raw_data_tfl_radar_raekoda.bag" doc="bag file to play" />
    <arg name="speed_limit"             default="40" />
    <arg name="tfl_detector"            default="camera"        doc="camera, mqtt, mqtt_camera_fusion" />
    <arg name="lanelet2_map_name"       default="tartu_large.osm" />
    <arg name="use_localization"        default="true"          doc="replace localization topics"/>
    <arg name="use_detection"           default="false"         doc="replace object detection topics"/>
    <arg name="use_planning"            default="true"          doc="replace planning topics"/>
    <arg name="use_control"             default="true"          doc="replace control topics"/>


    <!-- Use simulation time from bag -->
    <param name="/use_sim_time" value="true" />

    <!-- Play back bag -->
    <node pkg="rosbag" type="play" name="player" required="true" args="$(find common)/data/bags/$(arg bag_file) --clock --queue=1">
        <!-- Rename saved topics that are replaced by current ones -->
        <remap from="/tf" to="/bag/tf" />
        <remap from="/tf_static" to="/bag/tf_static" />
        <remap from="/localization/current_pose" to="/bag/localization/current_pose" if="$(arg use_localization)"/>
        <remap from="/localization/current_velocity" to="/bag/localization/current_velocity" if="$(arg use_localization)"/>
        <remap from="/localization/odometry" to="/bag/localization/odometry" if="$(arg use_localization)"/>
        <remap from="/detection/lidar/points_ground" to="/bag/detection/lidar/points_ground" if="$(arg use_detection)"/>
        <remap from="/detection/lidar/points_no_ground" to="/bag/detection/lidar/points_no_ground" if="$(arg use_detection)"/>
        <remap from="/detection/lidar/points_clustered" to="/bag/detection/lidar/points_clustered" if="$(arg use_detection)"/>
        <remap from="/detection/lidar/points_filtered" to="/bag/detection/lidar/points_filtered" if="$(arg use_detection)"/>
        <remap from="/detection/lidar/detected_objects" to="/bag/detection/lidar/detected_objects" if="$(arg use_detection)"/>
        <remap from="/detection/lidar/detected_objects_markers" to="/bag/detection/lidar/detected_objects_markers" if="$(arg use_detection)"/>
        <remap from="/detection/radar/detected_objects" to="/bag/detection/radar/detected_objects" if="$(arg use_detection)"/>
        <remap from="/detection/radar/detected_objects_markers" to="/bag/detection/radar/detected_objects_markers" if="$(arg use_detection)"/>
        <remap from="/detection/detected_objects" to="/bag/detection/detected_objects" if="$(arg use_detection)"/>
        <remap from="/detection/detected_objects_markers" to="/bag/detection/detected_objects_markers" if="$(arg use_detection)"/>
        <remap from="/detection/tracked_objects" to="/bag/detection/tracked_objects" if="$(arg use_detection)"/>
        <remap from="/detection/tracked_objects_markers" to="/bag/detection/tracked_objects_markers" if="$(arg use_detection)"/>
        <remap from="/detection/predicted_objects" to="/bag/detection/predicted_objects" if="$(arg use_detection)"/>
        <remap from="/detection/predicted_objects_markers" to="/bag/detection/predicted_objects_markers" if="$(arg use_detection)"/>
        <remap from="/detection/final_objects" to="/bag/detection/final_objects" if="$(arg use_detection)"/>
        <remap from="/detection/final_objects_markers" to="/bag/detection/final_objects_markers" if="$(arg use_detection)"/>
        <remap from="/planning/lanelet2_map_markers" to="/bag/planning/lanelet2_map_markers" if="$(arg use_planning)"/>
        <remap from="/planning/global_path" to="/bag/planning/global_path" if="$(arg use_planning)"/>
        <remap from="/planning/global_path_markers" to="/bag/planning/global_path_markers" if="$(arg use_planning)"/>
        <remap from="/planning/smoothed_path" to="/bag/planning/smoothed_path" if="$(arg use_planning)"/>
        <remap from="/planning/smoothed_path_markers" to="/bag/planning/smoothed_path_markers" if="$(arg use_planning)"/>
        <remap from="/planning/local_path" to="/bag/planning/local_path" if="$(arg use_planning)"/>
        <remap from="/planning/local_path_markers" to="/bag/planning/local_path_markers" if="$(arg use_planning)"/>
        <remap from="/planning/stop_line_markers" to="/bag/planning/stop_line_markers" if="$(arg use_planning)"/>
        <remap from="/planning/target_lane_markers" to="/bag/planning/target_lane_markers" if="$(arg use_planning)"/>
        <remap from="/detection/traffic_light_status" to="/bag/detection/traffic_light_status" />
        <remap from="/detection/camera_fl/traffic_light_status" to="/bag/detection/camera_fl/traffic_light_status" />
        <remap from="/detection/camera_fl/traffic_light_roi" to="/bag/detection/camera_fl/traffic_light_roi" />
        <remap from="/detection/camera_fr/traffic_light_status" to="/bag/detection/camera_fr/traffic_light_status" />
        <remap from="/detection/camera_fr/traffic_light_roi" to="/bag/detection/camera_fr/traffic_light_roi" />
        <remap from="/detection/camera/traffic_light_status" to="/bag/detection/camera/traffic_light_status" />
        <remap from="/detection/mqtt/traffic_light_status" to="/bag/detection/mqtt/traffic_light_status" />
        <remap from="/control/vehicle_cmd" to="/bag/control/vehicle_cmd" if="$(arg use_control)"/>
        <remap from="/dashboard/current_speed" to="/bag/dashboard/current_speed" />
        <remap from="/dashboard/target_speed" to="/bag/dashboard/target_speed" />
        <remap from="/dashboard/closest_object_distance" to="/bag/dashboard/closest_object_distance" />
        <remap from="/dashboard/closest_object_speed" to="closest_object_speed" />
        <remap from="/dashboard/acceleration" to="/bag/dashboard/acceleration" />
        <remap from="/dashboard/acceleration_limit" to="/bag/dashboard/acceleration_limit" />
        <remap from="/dashboard/deceleration_limit" to="/bag/dashboard/deceleration_limit" />
        <!-- For compatibility with older bags and Vella bags -->
        <remap from="/points_raw_center" to="/lidar_center/points_raw"/>
        <remap from="/points_raw_front" to="/lidar_front/points_raw"/>
        <remap from="/lidar_0/vlp32/pcl2" to="/lidar_center/points_raw"/>
    </node>


     <!-- Localization -->
    <include file="$(find practice_7)/launch/localization.launch" />
    
    <!-- Planning -->
    <include if="$(arg use_planning)" file="$(find practice_7)/launch/planning.launch">
        <arg name="lanelet2_map_name"   value="$(arg lanelet2_map_name)" />
        <arg name="speed_limit"         value="$(arg speed_limit)" />
    </include>

    <!-- Detection -->
    <include file="$(find practice_7)/launch/detection.launch" >
        <arg name="use_detection"       value="$(arg use_detection)" />
        <arg name="lanelet2_map_name"   value="$(arg lanelet2_map_name)" />
        <arg name="tfl_detector"        value="$(arg tfl_detector)" />
    </include>

    <!-- Control -->
    <include if="$(arg use_control)" file="$(find practice_7)/launch/control.launch" />

    <!-- Load vehicle config data -->
    <rosparam command="load" file="$(find practice_7)/config/vehicle.yaml"/>

    <!-- Load static transforms -->
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="vehicle_platform_frames">
        <remap from="robot_description" to="vehicle_platform_frames/robot_description" />
        <param name="robot_description" textfile="$(find practice_7)/config/platform.urdf" />
    </node>

    <include file="$(find practice_7)/launch/dashboard.launch" />

    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find practice_7)/rviz/practice_7.rviz" required="true" />


</launch>